File: src/main/java/io/vproxy/vfx/ui/shapes/VLine.java
package io.vproxy.vfx.ui.shapes;

import javafx.scene.Group;
import javafx.scene.Node;
import javafx.scene.paint.Paint;
import javafx.scene.shape.Circle;
import javafx.scene.shape.Line;
import javafx.scene.transform.Rotate;

public class VLine extends Group {
    private final Circle beginDot;
    private final Line line;
    private final Circle endDot;

    private final double width;
    private final double radius;

    private double startX;
    private double startY;
    private double endX;
    private double endY;
    private Paint fill;

    private Arrow arrowImageStart = null;
    private Arrow arrowImageEnd = null;

    public VLine(double width) {
        this.width = width;
        this.radius = width / 2;

        beginDot = new Circle(radius);
        beginDot.setStrokeWidth(0);
        line = new Line();
        line.setStrokeWidth(width);
        endDot = new Circle(radius);
        endDot.setStrokeWidth(0);

        getChildren().addAll(beginDot, line, endDot);
    }

    public void setStartStyle(EndpointStyle style) {
        if (style == EndpointStyle.ARROW) {
            var arrowImage = loadOrMakeArrowImage();
            calcDirection(arrowImage, false);
            arrowImage.setLayoutX(getStartX());
            arrowImage.setLayoutY(getStartY());
            arrowImageStart = arrowImage;
            getChildren().add(arrowImage);
        } else {
            if (arrowImageStart != null) {
                getChildren().remove(arrowImageStart);
                arrowImageStart = null;
            }
        }
    }

    public void setEndStyle(EndpointStyle style) {
        if (style == EndpointStyle.ARROW) {
            var arrowImage = loadOrMakeArrowImage();
            calcDirection(arrowImage, true);
            arrowImage.setLayoutX(getEndX());
            arrowImage.setLayoutY(getEndY());
            arrowImageEnd = arrowImage;
            getChildren().add(arrowImage);
        } else {
            if (arrowImageEnd != null) {
                getChildren().remove(arrowImageEnd);
                arrowImageEnd = null;
            }
        }
    }

    private void calcDirection(Node n, boolean toEnd) {
        double dx;
        double dy;
        if (toEnd) {
            dx = getEndX() - getStartX();
            dy = getEndY() - getStartY();
        } else {
            dx = getStartX() - getEndX();
            dy = getStartY() - getEndY();
        }
        if (dx == 0 && dy == 0) {
            return;
        }
        var l = Math.sqrt(dx * dx + dy * dy);
        var r = Math.acos(dx / l);
        if (dy < 0) {
            r = -r;
        }
        var d = r * 180 / Math.PI;
        var rotation = new Rotate(d);
        n.getTransforms().add(rotation);
    }

    private Arrow loadOrMakeArrowImage() {
        var arrow = new Arrow();
        arrow.setFill(fill);
        var ratio = width / 32;
        arrow.setScale(ratio);
        return arrow;
    }

    public void setStroke(Paint fill) {
        this.fill = fill;

        beginDot.setFill(fill);
        line.setStroke(fill);
        endDot.setFill(fill);
        if (arrowImageStart != null) {
            arrowImageStart.setFill(fill);
        }
        if (arrowImageEnd != null) {
            arrowImageEnd.setFill(fill);
        }
    }

    public double getStartX() {
        return startX;
    }

    public void setStartX(double startX) {
        this.startX = startX;
        beginDot.setLayoutX(startX);
        updateLine();
    }

    public double getStartY() {
        return startY;
    }

    public void setStartY(double startY) {
        this.startY = startY;
        beginDot.setLayoutY(startY);
        updateLine();
    }

    public double getEndX() {
        return endX;
    }

    public void setEndX(double endX) {
        this.endX = endX;
        endDot.setLayoutX(endX);
        updateLine();
    }

    public double getEndY() {
        return endY;
    }

    public void setEndY(double endY) {
        this.endY = endY;
        endDot.setLayoutY(endY);
        updateLine();
    }

    private void updateLine() {
        var dx = endX - startX;
        var dy = endY - startY;
        if (dx == 0 && dy == 0) {
            line.setVisible(false);
            return;
        }
        line.setVisible(true);
        var l = Math.sqrt(dx * dx + dy * dy);
        dx = dx * radius / l;
        dy = dy * radius / l;
        line.setStartX(startX + dx);
        line.setStartY(startY + dy);
        line.setEndX(endX - dx);
        line.setEndY(endY - dy);
        if (arrowImageStart != null) {
            arrowImageStart.setLayoutX(startX);
            arrowImageStart.setLayoutY(startY);
        }
        if (arrowImageEnd != null) {
            arrowImageEnd.setLayoutX(endX);
            arrowImageEnd.setLayoutY(endY);
        }
    }

    public void setStart(double x, double y) {
        setStartX(x);
        setStartY(y);
    }

    public void setEnd(double x, double y) {
        setEndX(x);
        setEndY(y);
    }
}

File: src-ts/src/ui/shapes/VLine.ts
import { Group } from '../../javafx/Pane.js';
import { Circle, LineShape } from '../../javafx/shapes.js';
import type { Node } from '../../javafx/Node.js';
import type { Paint } from '../../javafx/color.js';
import { Arrow } from './Arrow.js';
import { EndpointStyle } from './EndpointStyle.js';

export class VLine extends Group {
  private readonly beginDot: Circle;
  private readonly line: LineShape;
  private readonly endDot: Circle;

  private readonly width: number;
  private readonly radius: number;

  private startX = 0;
  private startY = 0;
  private endX = 0;
  private endY = 0;
  private fill: Paint | null = null;

  private arrowImageStart: Arrow | null = null;
  private arrowImageEnd: Arrow | null = null;

  constructor(width: number) {
    super();
    this.width = width;
    this.radius = width / 2;

    this.beginDot = new Circle(this.radius);
    this.beginDot.setStrokeWidth(0);
    this.line = new LineShape();
    this.line.setStrokeWidth(width);
    this.endDot = new Circle(this.radius);
    this.endDot.setStrokeWidth(0);

    // Order matters: line first, then dots on top. The line is drawn all
    // the way through the corner (see updateLine) so its rectangular end
    // doesn't leave a gap against the dot's curved edge — the dots then
    // cover the line's end-caps, matching the JavaFX visual outcome of
    // "union of shapes" without JavaFX's actual union rendering.
    this.getChildren().addAll(this.line, this.beginDot, this.endDot);
  }

  setStartStyle(style: EndpointStyle): void {
    if (style === EndpointStyle.ARROW) {
      const arrowImage = this.loadOrMakeArrowImage();
      this.calcDirection(arrowImage, false);
      arrowImage.setLayoutX(this.getStartX());
      arrowImage.setLayoutY(this.getStartY());
      this.arrowImageStart = arrowImage;
      this.getChildren().add(arrowImage);
    } else {
      if (this.arrowImageStart !== null) {
        this.getChildren().remove(this.arrowImageStart);
        this.arrowImageStart = null;
      }
    }
  }

  setEndStyle(style: EndpointStyle): void {
    if (style === EndpointStyle.ARROW) {
      const arrowImage = this.loadOrMakeArrowImage();
      this.calcDirection(arrowImage, true);
      arrowImage.setLayoutX(this.getEndX());
      arrowImage.setLayoutY(this.getEndY());
      this.arrowImageEnd = arrowImage;
      this.getChildren().add(arrowImage);
    } else {
      if (this.arrowImageEnd !== null) {
        this.getChildren().remove(this.arrowImageEnd);
        this.arrowImageEnd = null;
      }
    }
  }

  private calcDirection(n: Node, toEnd: boolean): void {
    let dx: number;
    let dy: number;
    if (toEnd) {
      dx = this.getEndX() - this.getStartX();
      dy = this.getEndY() - this.getStartY();
    } else {
      dx = this.getStartX() - this.getEndX();
      dy = this.getStartY() - this.getEndY();
    }
    if (dx === 0 && dy === 0) {
      return;
    }
    const l = Math.sqrt(dx * dx + dy * dy);
    const r = Math.acos(dx / l);
    const angle = dy < 0 ? -r : r;
    const d = (angle * 180) / Math.PI;
    n.el.style.transform = `rotate(${d}deg)`;
  }

  private loadOrMakeArrowImage(): Arrow {
    const arrow = new Arrow();
    arrow.setFill(this.fill);
    const ratio = this.width / 32;
    arrow.setScale(ratio);
    return arrow;
  }

  setStroke(fill: Paint): void {
    this.fill = fill;

    this.beginDot.setFill(fill);
    this.line.setStroke(fill);
    this.endDot.setFill(fill);
    if (this.arrowImageStart !== null) {
      this.arrowImageStart.setFill(fill);
    }
    if (this.arrowImageEnd !== null) {
      this.arrowImageEnd.setFill(fill);
    }
  }

  getStartX(): number { return this.startX; }

  setStartX(startX: number): void {
    this.startX = startX;
    this.beginDot.setLayoutX(startX);
    this.updateLine();
  }

  getStartY(): number { return this.startY; }

  setStartY(startY: number): void {
    this.startY = startY;
    this.beginDot.setLayoutY(startY);
    this.updateLine();
  }

  getEndX(): number { return this.endX; }

  setEndX(endX: number): void {
    this.endX = endX;
    this.endDot.setLayoutX(endX);
    this.updateLine();
  }

  getEndY(): number { return this.endY; }

  setEndY(endY: number): void {
    this.endY = endY;
    this.endDot.setLayoutY(endY);
    this.updateLine();
  }

  private updateLine(): void {
    const dx = this.endX - this.startX;
    const dy = this.endY - this.startY;
    if (dx === 0 && dy === 0) {
      this.line.setVisible(false);
      return;
    }
    this.line.setVisible(true);
    // Unlike the Java original, we do NOT offset the line endpoints by
    // `radius`. JavaFX unions overlapping shapes so the line can stop at
    // the dot's edge without leaving a gap; the DOM port renders each
    // shape independently and a perpendicular line end-cap meeting a
    // circular dot edge leaves a triangular gap (up to ~width/2 wide).
    // Instead we extend the line all the way through the corner and let
    // the dots (added after the line in the children list) sit on top to
    // cover the end-caps. Visual result matches JavaFX.
    this.line.setStartX(this.startX);
    this.line.setStartY(this.startY);
    this.line.setEndX(this.endX);
    this.line.setEndY(this.endY);
    if (this.arrowImageStart !== null) {
      this.arrowImageStart.setLayoutX(this.startX);
      this.arrowImageStart.setLayoutY(this.startY);
    }
    if (this.arrowImageEnd !== null) {
      this.arrowImageEnd.setLayoutX(this.endX);
      this.arrowImageEnd.setLayoutY(this.endY);
    }
  }

  setStart(x: number, y: number): void {
    this.setStartX(x);
    this.setStartY(y);
  }

  setEnd(x: number, y: number): void {
    this.setEndX(x);
    this.setEndY(y);
  }
}
